packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -147,33 +239,116 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
+        if (hubBalance >= minHubBalance + ethActualCharge) {
+            return;
+        }
+        uint256 depositAmount = targetHubBalance - hubBalance + ethActualCharge;
+        _refillHubDeposit(depositAmount);
+    }
+
+    function _refillHubDeposit(uint256 depositAmount) private {
+        uint256 balance = address(this).balance;
+        uint256 amountSwapped = 0;
+        if (balance < depositAmount) {
+            for (uint256 i = 0; i < tokens.length && balance + amountSwapped < depositAmount; i++) {
+                IERC20Metadata tokenIn = tokens[i];
+                uint256 tokenBalance = tokenIn.balanceOf(address(this));
+                if (tokenBalance > 0) {
+                    TokenSwapData memory tokenSwapData = tokensSwapData[tokenIn];
+                    uint256 quote = uint256(tokenSwapData.priceFeed.latestAnswer());
+                    uint256 amountOutMin = addSlippage(_tokenToWei(tokenBalance, tokenSwapData.priceDivisor, quote), tokenSwapData.slippage);
+                    if (amountOutMin < minSwapAmount) {
+                        continue;
+                    }
+                    uint256 amountOut = UniswapV3Helper.swapToToken(
+                        address(tokenIn),
+                        address(weth),
+                        tokenBalance,
+                        amountOutMin,
+                        tokenSwapData.uniswapPoolFee,
+                        uniswap
+                    );
+                    amountSwapped += amountOut;
+                }
+            }
+            if (amountSwapped > 0) {
+                UniswapV3Helper.unwrapWeth(uniswap, amountSwapped);
+            }
+        }
+        if (balance + amountSwapped > 0) {
+            relayHub.depositFor{value : balance + amountSwapped}(address(this));
+        }
+    }
+
+    function _withdrawToOwnerIfNeeded() private {
+        uint256 hubBalance = relayHub.balanceOf(address(this));
+        if (hubBalance >= minWithdrawalAmount + targetHubBalance) {
+        relayHub.withdraw(payable(owner()), hubBalance - targetHubBalance);
+        }
+    }
+
+    function _getTokenFromPaymasterData(bytes calldata paymasterData) internal pure returns (IERC20Metadata) {
+        return IERC20Metadata(address(bytes20(paymasterData[:20])));
+    }
+
+    function addPaymasterFee(uint256 charge) public view returns (uint256) {
+        return charge * (100 + paymasterFee) / 100;
+    }
+
+    function addSlippage(uint256 amount, uint8 slippage) public pure returns (uint256) {
+        return amount * (1000 - slippage) / 1000;
     }
 
     // as this Paymaster already has a permission from a user to operate the tokens on behalf of the gasless account,
     // it makes this same Paymaster a great recipient of a transaction if its only action is a pure token transfer
-    function transferToken(address target, uint256 value) external {
+    function transferToken(IERC20Metadata token, address target, uint256 value) external {
```

Comment:
This helper should be `transferAndCall(IERC20 token, uint value, address target, bytes callData)`
This makes integration with protocols such as "uniswap" simpler (otherwise, we need an explicit "UniswapPaymaster" to add just that)
It allows integration with any protocol that receives tokens, and performs a one-off method call (that is, doesn't depend about the sender of the tokens, which requires the protocol to be GSN-aware)